### ISSUE TITLE
is this normal GAN masks in FaceSwap ?

### ISSUE THREAD

**Issue by iperov** [shaoanlu/faceswap-GAN#50]
dots with oversharped image
is this normal , or faceswap developers adapted your plugin with errors?

![36072789-8173ea42-0f3f-11e8-8bc2-b2b8d8e9803b 1](https://user-images.githubusercontent.com/8076202/36072870-d54e5b1a-0f40-11e8-919d-51ea9a583b45.png)

**Comment by shaoanlu** [comment]
Jesus man, these masks are awesome. They are like dot arts. Please tell me how did you get it, seriously. What's the training data? How many iters. did you train on? I would really like to reproduce the result.

In terms of fakeswap, I did not look into the detail of the recent GAN update, but I believe its the same with my implementation. 

I have not seen such mask before, but I guess it might has something to do with lacking diversity of training data. However, the sharpness can be eliminated by GaussianBlur during video making so that's not a big deal in my opinion.

**Comment by iperov** [comment]
I just built https://github.com/deepfakes/faceswap latest release + 155 pull request.

I built windows version with cmd batches for myself, you can download it from here
test footage as on picture included in torrent
https://github.com/deepfakes/faceswap-playground/issues/39

**Comment by shaoanlu** [comment]
I've download the zip file through torrent. I only found 2 mp4 videos in the zip file, each < 30 sec.. So I think such masks are caused by too little training data. The model is thus heavily overfit to the training data such that one of preview mask column is purely white (which means the model believe it can reconstruct input image by 100% accuracy).

**Comment by iperov** [comment]
652 photos not enough for GAN ?
then how many?

**Comment by iperov** [comment]
actually white column is not purely white:
![2018-02-11_17-01-54](https://user-images.githubusercontent.com/8076202/36073651-8caf5f6e-0f4d-11e8-9610-569587e6ae84.gif)

**Comment by shaoanlu** [comment]
LOL, love this gif alot. It shows that what we see is not actual we see.

Are these 652 images come from the 2 videos (data_dst.mp4 and data_src.mp4)? If yes, then its definitely not enough since the extracted faces will look the same (under same lighting condition).

In my experiment, its better to have more than 1k image from various (>=3 at least) video source. And the extraction fps <5 so that there will not be too much duplicate images.

However, you can still try transfer learning if you have little training data. To be specific, we first train our model using other face dataset (e.g., Emma Watson, Donald Trump, celebA etc.). Once the pre-trained model shows good results (say, after 20k iters), then we train it using these 652 images.

But really, have you tried to make video using your current model? I'm imaging it produce faces with a lot of jitters. But considering how much face detail it generate in the previews, It might produces interesting results in video.

**Comment by iperov** [comment]
data_src.mp4 - 654 images

**Comment by shaoanlu** [comment]
Ahh that make sense. A 27-sec-long video with avg. 25fps. 27 x 25 is about 654. You probably want to add more images from other source.

**Comment by iperov** [comment]
result in current stage 
![2018-02-11_17-30-07](https://user-images.githubusercontent.com/8076202/36073907-73f721ec-0f51-11e8-8205-52c86539dc09.gif)

**Comment by shaoanlu** [comment]
I think the result can be much better if the face bounding box is more stable... i.e., the bbox position being a moving average of previous frames. As we can see, the bbox size and position in the gif above is not smoothed so it results in severe jitters. However, as far as I know there is no such functionality in faceswap.

**Comment by iperov** [comment]
so there is no point to train GAN model in https://github.com/deepfakes/faceswap ?

**Comment by shaoanlu** [comment]
The GAN (and non-GAN as well) will work if we have enough training data.

**Comment by iperov** [comment]
non-GAN works much better.

![2018-02-11_18-02-54](https://user-images.githubusercontent.com/8076202/36074193-fd6a2f42-0f55-11e8-9a1d-37cc69f217c0.gif)

So may be deepfakes/faceswap did something wrong when ported your model?

**Comment by shaoanlu** [comment]
I don't  think there are much difference between mine implementation and faceswap's. Perhaps GAN just not as data efficient as non-GAN.

**Comment by dfaker** [comment]
@shaoanlu  I think the GAN + perceptual loss is very promising, But I've never been convinced that the masking doesn't degrade down to edge detection. 

Have you considered training against a mask derived from the convex hull of the landmarks with the mask scaled down or multiplied by the rgb-face-loss*sigmoid to allow the network to reduce the mask only if it's beneficial to face reproduction?

**Comment by shaoanlu** [comment]
The landmarks will be detected no matter there is occlusion on faces or not. My understanding is that if we use convex hull mask as supervised ground truth, the model will overfit so that it can not handle occlusion anymore. On the contrast, I believe the mask is able to learn more or less semantic segmentation, as shown in [LR-GAN](https://arxiv.org/abs/1703.01560). But we have to find better architecture and loss function for face swapping.

**Comment by dfaker** [comment]
> The landmarks will be detected no matter there is occlusion on faces or not.

Yeah hence the need to give the network a 'loophole' by biasing the loss using rgb-face-loss*sigmoid, if the face is being generated well the loss will be near zero and we won't care about adjustments to the masking layer, if the generation is poor however the rgb-face-loss*sigmoid is high and we do rigidly enforce the ground truth mask.

**Comment by shaoanlu** [comment]
Can you elaborate more on the rgb-face-loss*sigmoid?

**Comment by dfaker** [comment]
So we have the rgb-face-loss which represents purely the accuracy of the reproduction of the  unmasked face.

The output of that is simply passed through some mapping function, perhaps sigmoid perhaps something harder, to force high losses towards 0 and low losses towards 1, then it becomes suitable for multiplication with the mask loss enforcing the full loss against the ground truth mask when the face is accurate but allowing the mask to take on any value when the face is innacurate.

If an occlusion appears it returns a high rgb-face-loss as an unexpected feature to appear in a face, this is forced low by the rgb-face-loss*sigmoid which then when multiplied with the mask loss it biases it towards zero which allows the mask output to move away from the ground-truth-mask.

For normal areas of the face without occlusion, rgb-face-loss is low which is forced high, passing the full loss back to the mask layer.

**Comment by shaoanlu** [comment]
I think this is viable. It would be good if someone can run some experiment for prove of concept.

**Comment by shaoanlu** [comment]
Here are the results I got after training for ~10k iters (about 4 hrs on K80) with perceptual loss. The masked faces are fairly good. So what I've learned is that, as long as perceptual loss is introduced during training, we probably don't have to prepare training data as much as I though it should.

![0218_masked_pl](https://user-images.githubusercontent.com/22006284/36354834-67e85340-151d-11e8-8f8f-c8ca77da2834.png)
![0218_mask_pl](https://user-images.githubusercontent.com/22006284/36354838-68532332-151d-11e8-99cf-21a6828b38fc.png)

However, the generated mask can not deal with hard sample, e.g., faces with motion blurs at the bottom left of figures above. As a result, the output video quality on non-stationary face is sub-optimal. 

![shia_downey](https://user-images.githubusercontent.com/22006284/36354901-49164778-151e-11e8-8ab8-2c00c60ff05d.gif)

We can see there is one quick moment that it failed to transform the face to Downey.

**Comment by iperov** [comment]
how I can get such masks in faceswap ?

**Comment by shaoanlu** [comment]
Yet another results trained for 15k **without** perceptual loss. The model is not exactly the same with v2 model but the modifications should have little impact on output quality theoretically.

![just_do_it_masked_wopl](https://user-images.githubusercontent.com/22006284/36380351-4bb62a28-15c5-11e8-8099-c44920aa4eda.png)
![just_do_it_mask_wopl](https://user-images.githubusercontent.com/22006284/36380352-4c3f5c58-15c5-11e8-8391-dff193935688.png)

In the light of my recent experiments, I think the sharp masks shown in your post are due to too long of training so that the model is heavily overfit.

**Comment by Apollo122** [comment]
Is perceptional loss implemented in faceswap?

**Comment by iperov** [comment]
no

**Comment by MisterGenerosity** [comment]
Hey shaoanlu, I noticed you mentioned color correction above

> But it is somehow eliminated after color correction.

What do you do to produce color correction? Meaning, what software or, perhaps, what code? As I'm getting into this I'm finding my subject faces have such different skin tones, and I'm concerned it'll look like rubbish when I'm done.

**Comment by shaoanlu** [comment]
Color correction is implemented through histogram matching. You can find the piece of code in v2_test_video notebook.

The reason I did not include perceptual loss into faceswap is because it requires keras >2.1.1 (and more VRAM). I worried that it will lead to lots of issues posts on the repo., thus I decide to remove it.

**Comment by iperov** [comment]
is histogram matching implemented in faceswap?

**Comment by shaoanlu** [comment]
Probably not, I added it recently. I did not follow updates of faceswap closely so maybe there are some similar works done.

**Comment by Apollo122** [comment]
Nope. Repo is pretty idle now.

**Comment by MonkeyNi** [comment]
![log_280k_black_backg_tom](https://user-images.githubusercontent.com/37827378/49358484-cb2a9080-f70d-11e8-906d-7d748a3bf5ec.png)

I trained faceswap GAN model and got the showed result. FaceA and FaceB are about 1300+ images which were extracted from videos. I think the mask and preview result are good. But the swapped result is not satisfactory. 
![1](https://user-images.githubusercontent.com/37827378/49358720-a4208e80-f70e-11e8-985f-b0b65e85f56d.png)

It is not like Tom Cruise at all...
Please help me out, thanks!

**Comment by mrgloom** [comment]
About checkerboard artifacts
https://distill.pub/2016/deconv-checkerboard/
